File: GitToolBox/src/main/java/zielu/gittoolbox/config/GitToolBoxConfig2.java
package zielu.gittoolbox.config;

import com.google.common.collect.Lists;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.components.PersistentStateComponent;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import com.intellij.util.xmlb.XmlSerializerUtil;
import com.intellij.util.xmlb.annotations.Transient;
import java.util.List;
import java.util.stream.Collectors;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import zielu.gittoolbox.ui.StatusPresenter;
import zielu.gittoolbox.ui.StatusPresenters;
import zielu.gittoolbox.ui.update.DefaultUpdateProjectAction;

@State(
    name = "GitToolBoxAppSettings2",
    storages = @Storage("git_toolbox_2.xml")
)
public class GitToolBoxConfig2 implements PersistentStateComponent<GitToolBoxConfig2> {
  public String presentationMode = StatusPresenters.arrows.key();
  public boolean behindTracker = true;
  public boolean showStatusWidget = true;
  public boolean showProjectViewStatus = true;
  public boolean showBlameWidget = true;
  public boolean showEditorInlineBlame = true;
  public String updateProjectActionId = DefaultUpdateProjectAction.ID;
  public List<DecorationPartConfig> decorationParts = Lists.newArrayList(
      DecorationPartConfig.builder().withType(DecorationPartType.LOCATION).withPrefix("- ").build(),
      DecorationPartConfig.builder().withType(DecorationPartType.BRANCH).build(),
      DecorationPartConfig.builder().withType(DecorationPartType.STATUS).build(),
      DecorationPartConfig.builder().withType(DecorationPartType.TAGS_ON_HEAD).withPrefix("(").withPostfix(")").build(),
      DecorationPartConfig.builder().withType(DecorationPartType.CHANGED_COUNT).withPrefix("/ ").build()
  );
  public CommitCompletionMode commitDialogCompletionMode = CommitCompletionMode.AUTOMATIC;
  public AuthorNameType blameInlineAuthorNameType = AuthorNameType.LASTNAME;
  public AuthorNameType blameStatusAuthorNameType = AuthorNameType.LASTNAME;
  public DateType blameInlineDateType = DateType.AUTO;
  public boolean blameInlineShowSubject = true;
  public AbsoluteDateTimeStyle absoluteDateTimeStyle = AbsoluteDateTimeStyle.FROM_LOCALE;
  public boolean showChangesInStatusBar = true;

  public boolean previousVersionMigrated;

  public static GitToolBoxConfig2 getInstance() {
    return ServiceManager.getService(GitToolBoxConfig2.class);
  }


  public GitToolBoxConfig2 copy() {
    GitToolBoxConfig2 copy = new GitToolBoxConfig2();
    copy.presentationMode = presentationMode;
    copy.behindTracker = behindTracker;
    copy.showStatusWidget = showStatusWidget;
    copy.showProjectViewStatus = showProjectViewStatus;
    copy.showBlameWidget = showBlameWidget;
    copy.showEditorInlineBlame = showEditorInlineBlame;
    copy.updateProjectActionId = updateProjectActionId;
    copy.decorationParts = decorationParts.stream().map(DecorationPartConfig::copy).collect(Collectors.toList());
    copy.commitDialogCompletionMode = commitDialogCompletionMode;
    copy.blameInlineAuthorNameType = blameInlineAuthorNameType;
    copy.blameStatusAuthorNameType = blameStatusAuthorNameType;
    copy.blameInlineDateType = blameInlineDateType;
    copy.blameInlineShowSubject = blameInlineShowSubject;
    copy.absoluteDateTimeStyle = absoluteDateTimeStyle;
    copy.showChangesInStatusBar = showChangesInStatusBar;

    copy.previousVersionMigrated = previousVersionMigrated;
    return copy;
  }

  @Transient
  public StatusPresenter getPresenter() {
    return StatusPresenters.forKey(presentationMode);
  }

  public void setPresenter(StatusPresenter presenter) {
    presentationMode = presenter.key();
  }

  public boolean isPresenterChanged(StatusPresenter presenter) {
    return !presentationMode.equals(presenter.key());
  }

  public boolean isShowStatusWidgetChanged(boolean showStatusWidget) {
    return this.showStatusWidget != showStatusWidget;
  }

  public boolean isShowProjectViewStatusChanged(boolean showProjectViewStatus) {
    return this.showProjectViewStatus != showProjectViewStatus;
  }

  public boolean isBehindTrackerChanged(boolean behindTracker) {
    return this.behindTracker != behindTracker;
  }

  public String getUpdateProjectActionId() {
    return updateProjectActionId;
  }

  public void setUpdateProjectActionId(String id) {
    updateProjectActionId = id;
  }

  public boolean isUpdateProjectActionId(@NotNull String id) {
    return !updateProjectActionId.equals(id);
  }

  public boolean isDecorationPartsChanged(List<DecorationPartConfig> decorationParts) {
    return !this.decorationParts.equals(decorationParts);
  }

  public boolean isShowBlameWidgetChanged(boolean showBlameWidget) {
    return this.showBlameWidget != showBlameWidget;
  }

  public boolean isShowEditorInlineBlameChanged(boolean showEditorInlineBlame) {
    return this.showEditorInlineBlame != showEditorInlineBlame;
  }

  public boolean isCommitDialogCompletionModeChanged(CommitCompletionMode commitDialogCompletionMode) {
    return this.commitDialogCompletionMode != commitDialogCompletionMode;
  }

  public boolean isBlameInlineAuthorNameTypeChanged(AuthorNameType blameAuthorNameType) {
    return this.blameInlineAuthorNameType != blameAuthorNameType;
  }

  public boolean isBlameInlineDateTypeChanged(DateType blameDateType) {
    return this.blameInlineDateType != blameDateType;
  }

  public boolean isBlameInlineShowSubjectChanged(boolean blameInlineShowSubject) {
    return this.blameInlineShowSubject != blameInlineShowSubject;
  }

  public boolean isBlameInlinePresentationChanged(GitToolBoxConfig2 other) {
    return isBlameInlineAuthorNameTypeChanged(other.blameInlineAuthorNameType)
        || isBlameInlineDateTypeChanged(other.blameInlineDateType)
        || isBlameInlineShowSubjectChanged(other.blameInlineShowSubject)
        || isAbsoluteDateTimeStyleChanged(other.absoluteDateTimeStyle);
  }

  public boolean isBlameStatusAuthorNameTypeChanged(AuthorNameType blameAuthorNameType) {
    return this.blameStatusAuthorNameType != blameAuthorNameType;
  }

  public boolean isBlameStatusPresentationChanged(GitToolBoxConfig2 other) {
    return isBlameStatusAuthorNameTypeChanged(other.blameStatusAuthorNameType)
        || isAbsoluteDateTimeStyleChanged(other.absoluteDateTimeStyle);
  }

  public boolean isAbsoluteDateTimeStyleChanged(AbsoluteDateTimeStyle other) {
    return absoluteDateTimeStyle != other;
  }

  public boolean isShowChangesInStatusBarChanged(boolean trackChanges) {
    return this.showChangesInStatusBar != trackChanges;
  }

  @Transient
  public boolean isChangesTrackingEnabled() {
    return showChangesInStatusBar || decorationParts.stream()
        .anyMatch(part -> part.type == DecorationPartType.CHANGED_COUNT);
  }

  @Transient
  public boolean isStatusBarWidgetVisible() {
    return showStatusWidget || isChangesTrackingEnabled();
  }

  @Nullable
  @Override
  public GitToolBoxConfig2 getState() {
    return this;
  }

  public void fireChanged(@NotNull GitToolBoxConfig2 previousConfig) {
    ApplicationManager.getApplication().getMessageBus().syncPublisher(ConfigNotifier.CONFIG_TOPIC)
        .configChanged(previousConfig, this);
  }

  @Override
  public void loadState(@NotNull GitToolBoxConfig2 state) {
    XmlSerializerUtil.copyBean(state, this);
  }
}

File: GitToolBox/src/main/kotlin/zielu/gittoolbox/config/GitToolBoxConfig2.kt
package zielu.gittoolbox.config

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.PersistentStateComponent
import com.intellij.openapi.components.State
import com.intellij.openapi.components.Storage
import com.intellij.util.xmlb.XmlSerializerUtil
import com.intellij.util.xmlb.annotations.Transient
import zielu.gittoolbox.extension.update.UpdateProjectAction
import zielu.gittoolbox.ui.StatusPresenter
import zielu.gittoolbox.ui.StatusPresenters
import zielu.gittoolbox.ui.update.DefaultUpdateProjectAction
import zielu.gittoolbox.ui.update.UpdateProjectActionService
import zielu.gittoolbox.util.AppUtil

@State(name = "GitToolBoxAppSettings2", storages = [Storage("git_toolbox_2.xml")])
internal data class GitToolBoxConfig2(
  var presentationMode: String = StatusPresenters.arrows.key(),
  var behindTracker: Boolean = true,
  var showStatusWidget: Boolean = true,
  var showProjectViewStatus: Boolean = true,
  var showBlameWidget: Boolean = true,
  var showEditorInlineBlame: Boolean = true,
  var updateProjectActionId: String = DefaultUpdateProjectAction.ID,
  var commitDialogCompletionMode: CommitCompletionMode = CommitCompletionMode.AUTOMATIC,
  var blameInlineAuthorNameType: AuthorNameType = AuthorNameType.LASTNAME,
  var blameStatusAuthorNameType: AuthorNameType = AuthorNameType.LASTNAME,
  var blameInlineDateType: DateType = DateType.AUTO,
  var blameInlineShowSubject: Boolean = true,
  var absoluteDateTimeStyle: AbsoluteDateTimeStyle = AbsoluteDateTimeStyle.FROM_LOCALE,
  var showChangesInStatusBar: Boolean = true,
  var previousVersionMigrated: Boolean = false,
  var decorationParts: List<DecorationPartConfig> = arrayListOf(
    DecorationPartConfig.builder().withType(DecorationPartType.LOCATION)
      .withPrefix("- ")
      .build(),
    DecorationPartConfig.builder().withType(DecorationPartType.BRANCH).build(),
    DecorationPartConfig.builder().withType(DecorationPartType.STATUS).build(),
    DecorationPartConfig.builder().withType(DecorationPartType.TAGS_ON_HEAD)
      .withPrefix("(")
      .withPostfix(")")
      .build(),
    DecorationPartConfig.builder().withType(DecorationPartType.CHANGED_COUNT)
      .withPrefix("/ ")
      .build()
  ),
  var extrasConfig: ExtrasConfig = ExtrasConfig()
) : PersistentStateComponent<GitToolBoxConfig2> {

  companion object {
    @JvmStatic
    fun getInstance(): GitToolBoxConfig2 {
      return AppUtil.getServiceInstance(GitToolBoxConfig2::class.java)
    }
  }

  @Transient
  fun copy(): GitToolBoxConfig2 {
    return GitToolBoxConfig2(
      presentationMode,
      behindTracker,
      showStatusWidget,
      showProjectViewStatus,
      showBlameWidget,
      showEditorInlineBlame,
      updateProjectActionId,
      commitDialogCompletionMode,
      blameInlineAuthorNameType,
      blameStatusAuthorNameType,
      blameInlineDateType,
      blameInlineShowSubject,
      absoluteDateTimeStyle,
      showChangesInStatusBar,
      previousVersionMigrated,
      decorationParts.map { it.copy() },
      extrasConfig.copy()
    )
  }

  @Transient
  fun getPresenter(): StatusPresenter {
    return StatusPresenters.forKey(presentationMode)
  }

  fun setPresenter(presenter: StatusPresenter) {
    presentationMode = presenter.key()
  }

  @Transient
  fun getUpdateProjectAction(): UpdateProjectAction {
    return UpdateProjectActionService.getInstance().getById(updateProjectActionId)
  }

  fun setUpdateProjectAction(action: UpdateProjectAction) {
    updateProjectActionId = action.getId()
  }

  fun isBlameInlinePresentationChanged(other: GitToolBoxConfig2): Boolean {
    return blameInlineAuthorNameType != other.blameInlineAuthorNameType ||
      blameInlineDateType != other.blameInlineDateType ||
      blameInlineShowSubject != other.blameInlineShowSubject ||
      absoluteDateTimeStyle != other.absoluteDateTimeStyle
  }

  fun isBlameStatusPresentationChanged(other: GitToolBoxConfig2): Boolean {
    return blameStatusAuthorNameType != other.blameStatusAuthorNameType ||
      absoluteDateTimeStyle != other.absoluteDateTimeStyle
  }

  @Transient
  fun isChangesTrackingEnabled(): Boolean {
    return showChangesInStatusBar || decorationParts.stream()
      .anyMatch { part: DecorationPartConfig -> part.type == DecorationPartType.CHANGED_COUNT }
  }

  fun fireChanged(previousConfig: GitToolBoxConfig2) {
    ApplicationManager.getApplication().messageBus
      .syncPublisher(AppConfigNotifier.CONFIG_TOPIC)
      .configChanged(previousConfig, this)
  }

  override fun getState(): GitToolBoxConfig2 = this

  override fun loadState(state: GitToolBoxConfig2) {
    XmlSerializerUtil.copyBean(state, this)
  }
}
